packages/consumption/src/modules/attributes/AttributesController.ts

```diff
@@ -425,10 +433,59 @@ export class AttributesController extends ConsumptionBaseController {
         return { predecessor, successor };
     }
 
+    private async succeedChildrenOfComplexAttribute(parentSuccessorId: CoreId) {
+        const parentSuccessor = (await this.getLocalAttribute(parentSuccessorId))!;
+        const childAttributeValues: AbstractAttributeValue[] = Object.values(parentSuccessor.content.value).filter((elem) => elem instanceof AbstractAttributeValue);
+
+        for (const childAttributeValue of childAttributeValues) {
+            let currentParent = await this.getLocalAttribute(parentSuccessorId);
+
+            let child;
+            while (typeof child === "undefined" && currentParent?.succeeds) {
+                const currentPredecessor = (await this.getLocalAttribute(currentParent.succeeds))!;
+                currentParent = currentPredecessor;
+                child = await this.getChildAttributesByValueType(currentParent.id, childAttributeValue.constructor);
+            }
+
+            const childPredecessorId = child?.id;
+            if (typeof childPredecessorId !== "undefined") {
+                await this._succeedAttributeUnsafe(childPredecessorId, {
+                    content: IdentityAttribute.from({
+                        value: childAttributeValue.toJSON() as AttributeValues.Identity.Json,
+                        owner: this.identity.address
+                    }),
+                    parentId: parentSuccessorId,
+                    createdAt: parentSuccessor.createdAt
+                });
+            } else {
+                await this.createAttributeUnsafe({
+                    content: IdentityAttribute.from({
+                        value: childAttributeValue.toJSON() as AttributeValues.Identity.Json,
+                        owner: this.identity.address
+                    }),
+                    parentId: parentSuccessorId,
+                    createdAt: parentSuccessor.createdAt
+                });
+            }
+        }
+    }
```

Comment:
Good catch, the ! operator should only be used in the tests.

Reply:
This function is called only immediately after the parent attribute has been created and thus the `parentSuccessor` attribute is guaranteed to exist. We deliberately removed the check for this very reason (although we did think about the pros and cons of doing so).

I reckon we just add the checks to make the method self-contained (and to avoid further